src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,216 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Razor;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer
+{
+    /// <summary>
+    /// Caches tokens on a per-line basis for documents based on semantic version.
+    /// </summary>
+    /// <remarks>
+    /// The cache makes the assumption that the tokens passed in are complete for the given range.
+    /// </remarks>
+    internal sealed class SemanticTokensCache
+    {
+        private const int MaxSemanticVersionPerDoc = 4;
+
+        // Nested cache mapping (URI -> (semanticVersion -> (line #s -> tokens on line)))
+        private readonly MemoryCache<DocumentUri, MemoryCache<VersionStamp, Dictionary<int, IReadOnlyList<int>>>> _cache = new();
+
+        /// <summary>
+        /// Caches tokens on a per-line basis. If the given line has already been cached for the document
+        /// and its semantic version, the line's tokens will be ignored.
+        /// </summary>
+        /// <param name="uri">The URI associated with the passed-in tokens.</param>
+        /// <param name="semanticVersion">The document's semantic version associated with the passed-in tokens.</param>
+        /// <param name="range">The range associated with the passed-in tokens.</param>
+        /// <param name="tokens">The complete set of tokens for the passed-in range.</param>
+        internal void CacheTokens(
```

Comment:
public plz (to match rest of codebase).  it has the same effect, but we tend to just keep methods public/private/protected in an internal type, and we only use 'internal' to indicate "this is in a public type, but we don't want other people to use it".